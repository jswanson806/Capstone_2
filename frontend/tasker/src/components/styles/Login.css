.login-container {
  display: flex;
  align-items: stretch;
  height: 100vh; /* Make the login container full height */
}

.login-message-container,
.login-form-container {
  flex: 1; /* Each container takes up half of the available space */
}

.login-message-container {
  display: block;
  background-color: #040404; /* Set a background color for the image container */
  padding: 20px;
  padding-bottom: 40px
}

.login-message-container p {
  display: relative;
  font-size: 60px;
  color: whitesmoke; /* Ensure the image doesn't exceed its container */
  margin-bottom: auto;

}

.login-form-container {
  display: flex;
  flex-direction: column;
  text-align: center;
  justify-content: center;
  align-items: center;
  padding: 20px; /* Add padding for spacing */
  margin-bottom: 25vh;
}

.login-button-container {
  display: flex;
  justify-content: center;
}

.login-signup-message-container {
  text-align: center;
  padding-top: 15px;
}

.login-signup-message-container p {
  margin-bottom: auto;
}

.login-header {
  display: flex;
  justify-content: center;
  margin-bottom: 10px;
}

/* Media query for smaller screens */
@media screen and (max-width: 768px) {
  .login-container {
    flex-direction: column; /* Stack elements vertically on smaller screens */
  }

  .login-form-container {
    flex: 1; /* Take up the entire width of the screen */
    padding: 20px; /* Add padding for spacing */
  }
}