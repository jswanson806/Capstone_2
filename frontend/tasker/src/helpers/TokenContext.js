import React, { createContext, useState } from 'react';

// Create the context
const TokenContext = createContext();

// Create a provider for the context
const TokenProvider = ({ children }) => {
  const [token, setToken] = useState(localStorage.getItem('token') || '');

  // Function to update the token in the state and localStorage
  const updateToken = (newToken) => {
    setToken(newToken);
    localStorage.setItem('token', newToken);
  };

  return (
    <TokenContext.Provider value={{ token, updateToken }}>
      {children}
    </TokenContext.Provider>
  );
};

export { TokenContext, TokenProvider };
