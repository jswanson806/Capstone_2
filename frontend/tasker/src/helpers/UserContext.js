import React, { createContext, useState } from 'react';

// Create the context
const UserContext = createContext();

// Create a provider for the context
const UserProvider = ({ children }) => {
  const [user, setUser] = useState(localStorage.getItem('user') || '');

  // Function to update the user in the state and localStorage
  const updateUser = (newUser) => {
    setUser(newUser);
    localStorage.setItem('user', newUser);
  };

  return (
    <UserContext.Provider value={{ user, updateUser }}>
      {children}
    </UserContext.Provider>
  );
};

export { UserContext, UserProvider };
