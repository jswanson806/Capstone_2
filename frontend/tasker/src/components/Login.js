import React, { useState, useContext } from "react";
import { TokenContext } from '../helpers/TokenContext.js';
import { useNavigate } from "react-router-dom";
import TaskerApi from "../api.js";
import { UserContext } from "../helpers/UserContext.js";
import {Button, Form, FormGroup, FormText, Input, Label} from "reactstrap";
import "./styles/Login.css";


/**
 * Login component for user authentication.
 *
 * This component renders a login form where users can enter their email and password
 * to authenticate themselves. The user is then redirected to the
 * dashboard.
 *
 * @returns {JSX.Element} The rendered login form component.
 */

const Login = () => {

    const navigate = useNavigate();

    const { updateToken } = useContext(TokenContext);
    const { updateUser } = useContext(UserContext);

    const INITIAL_STATE = {
        email: '',
        password: '',
    }

    const [formData, setFormData] = useState(INITIAL_STATE);

    /** handles events on the form to update as the user types */
    const handleChange = (e) => {
        const {name, value} = e.target;
        setFormData(formData => ({
            ...formData,
            [name]: value
        }))
    }
    /** handles submission of the form 
     * 
     * Calls api to log user in
     * Updates token and user context state
     * 
     * resets the form data
     * 
     * redirects user to jobs page
    */
    const handleSubmit = async (e) => {
        e.preventDefault();

        const userInfo = {
            email: formData.email,
            password: formData.password,
        }

        try {
            // call api to login user
            const res = await TaskerApi.login(userInfo.email, userInfo.password);
            // destructure token and user objects from response
            const { token, user } = res;
            // create stringified user object to store in local storage
            const newUser = JSON.stringify({id: user.id, email: user.email, isWorker: user.isWorker});
            
            // update localStorage token and user
            updateToken(token);
            updateUser(newUser);

            // clear the form data
            setFormData(INITIAL_STATE);
            // redirect to jobs
            navigate("/jobs");
        } catch(err) {
            return err;
        }
    }


    return (
        <div className="login-container">
            <div className="login-message-container">
                <p>Welcome Back to Tasker!</p>
            </div>
            
            <div className="login-form-container">
            
                <div className="login-header">
                    <h3>Login</h3>
                </div>
                
                <Form onSubmit={handleSubmit}>
                    <FormGroup>
                    <Label htmlFor="email">Email</Label>
                    <Input
                        id="email"
                        type="text"
                        name="email"
                        placeholder="Email Address"
                        data-testid="login-form-email-input"
                        value={FormData.email}
                        onChange={handleChange}
                    />
                    </FormGroup>
                    <FormGroup>
                    <Label htmlFor="password">Password</Label>
                    <Input
                        id="password"
                        type="password"
                        name="password"
                        placeholder="********"
                        data-testid="login-form-password-input"
                        value={FormData.password}
                        onChange={handleChange}
                    />
                    </FormGroup>
                    <div className="login-button-container">
                        <Button className="login-button" type="submit" color="info" data-testid="login-form-button">Login</Button>
                    </div>
                    <div className="login-signup-message-container">
                        <p>Not yet registered?</p>
                        <a href="/signup">Sign Up</a>
                    </div>
                </Form>
            </div>
        </div>
    );
}

export default Login;