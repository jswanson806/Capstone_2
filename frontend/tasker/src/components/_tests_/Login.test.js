import React from 'react';
import {act, render, screen, fireEvent} from '@testing-library/react';
import { MemoryRouter } from "react-router-dom";
import { TokenContext } from '../../helpers/TokenContext.js';
import { UserContext } from '../../helpers/UserContext.js';
import TaskerApi from '../../api.js';
import Login from '../Login.js';

const userValue = { user: '{"id": 1, "email": "test@email.com", "isWorker": true}' };

describe("login form smoke and snapshot tests", () => {

    it("should render without crashing", async () => {

        await act(async () => {render(
            <MemoryRouter>
                <TokenContext.Provider value={'dummyToken'}>
                    <UserContext.Provider value={userValue}>
                        <Login />
                    </UserContext.Provider>
                </TokenContext.Provider>
            </MemoryRouter>
        )
        });
    });

    it("should match snapshot", async() => {
        let asFragment;

        await act(async () => {const {asFragment: fragment} = render(
            <MemoryRouter>
                <TokenContext.Provider value={'dummyToken'}>
                    <UserContext.Provider value={userValue}>
                        <Login />
                    </UserContext.Provider>
                </TokenContext.Provider>
            </MemoryRouter>
        )
        asFragment = fragment;
        });

        expect(asFragment()).toMatchSnapshot();
    })
})

describe("Handles form input correctly", () => {

    it('updates form input correctly', async () => {
        await act(async () => {render(
            <MemoryRouter>
                <TokenContext.Provider value={'dummyToken'}>
                    <UserContext.Provider value={userValue}>
                        <Login />
                    </UserContext.Provider>
                </TokenContext.Provider>
            </MemoryRouter>
        )
        });

        const emailInput = screen.getByTestId("login-form-email-input");
        const password = screen.getByTestId("login-form-password-input");
        fireEvent.change(emailInput, {target: {value: "test@email.com"}});
        fireEvent.change(password, {target: {value: "password"}});

        expect(emailInput).toHaveValue("test@email.com");
        expect(password).toHaveValue("password");

    })

    it('should handle form submission', async () => {

        const mockLogin = jest.spyOn(TaskerApi, 'login')

        await act(async () => {render(
            <MemoryRouter>
                <TokenContext.Provider value={'dummyToken'}>
                    <UserContext.Provider value={userValue}>
                        <Login />
                    </UserContext.Provider>
                </TokenContext.Provider>
            </MemoryRouter>
        )
        });

        const emailInput = screen.getByTestId("login-form-email-input");
        const password = screen.getByTestId("login-form-password-input");
        fireEvent.change(emailInput, {target: {value: "test@email.com"}});
        fireEvent.change(password, {target: {value: "password"}});

        const btn = screen.getByTestId("login-form-button");

        await act(async () => {
            fireEvent.click(btn);
        })

        expect(mockLogin).toHaveBeenCalledWith("test@email.com", "password")

    })

})