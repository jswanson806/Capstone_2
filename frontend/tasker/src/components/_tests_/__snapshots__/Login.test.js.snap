// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`login form smoke and snapshot tests should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="login-container"
  >
    <div
      class="login-centered-card"
    >
      <h5>
        Login
      </h5>
      <form
        class=""
      >
        <small
          class="form-text text-muted"
          for="email"
        >
          Email
        </small>
        <input
          data-testid="login-form-email-input"
          id="email"
          name="email"
          placeholder="Email Address"
          type="text"
          value=""
        />
        <small
          class="form-text text-muted"
          for="password"
        >
          Password
        </small>
        <input
          data-testid="login-form-password-input"
          id="password"
          name="password"
          placeholder="********"
          type="password"
          value=""
        />
        <button
          class="login-button btn btn-info"
          data-testid="login-form-button"
          type="submit"
        >
          Login
        </button>
      </form>
    </div>
  </div>
</DocumentFragment>
`;
